crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -216,74 +216,139 @@ impl<'db> InnerIntersectionBuilder<'db> {
-        match ty {
-            Type::Intersection(inter) => {
-                let pos = inter.positive(db);
-                let neg = inter.negative(db);
-                self.positive.extend(pos.difference(&self.negative));
-                self.negative.extend(neg.difference(&self.positive));
-                self.positive.retain(|elem| !neg.contains(elem));
-                self.negative.retain(|elem| !pos.contains(elem));
+        if let Type::Intersection(other) = new_positive {
+            for pos in other.positive(db) {
+                self.add_positive(db, *pos);
             }
-            _ => {
-                if !self.negative.remove(&ty) {
-                    self.positive.insert(ty);
-                };
+            for neg in other.negative(db) {
+                self.add_negative(db, *neg);
+            }
+        } else {
+            // ~Literal[True] & bool = Literal[False]
+            if let Type::Instance(class_type) = new_positive {
+                if class_type.is_known(db, KnownClass::Bool) {
+                    let mut found_bool_literal = None;
+                    for neg in &self.negative {
+                        if let Type::BooleanLiteral(bool) = neg {
+                            found_bool_literal = Some(*bool);
+                        }
+                    }
+
+                    if let Some(bool) = found_bool_literal {
+                        *self = Self::new();
+                        self.positive.insert(Type::BooleanLiteral(!bool));
+                        return;
+                    }
+                }
+            }
+
+            let mut to_remove = None;
+            for (index, existing_positive) in self.positive.iter().enumerate() {
+                // S & T = S    if S <: T
+                if existing_positive.is_subtype_of(db, new_positive) {
+                    return;
+                }
+                // same rule, reverse order
+                if new_positive.is_subtype_of(db, *existing_positive) {
+                    to_remove = Some(index);
+                }
+                // A & B = Never    if A and B are disjoint
+                if new_positive.is_disjoint_from(db, *existing_positive) {
+                    *self = Self::new();
+                    return;
+                }
             }
+            if let Some(index) = to_remove {
+                self.positive.remove_index(index);
+            }
+
+            let mut to_remove = None;
+            for (index, existing_negative) in self.negative.iter().enumerate() {
+                // S & ~T = Never    if S <: T
+                if new_positive.is_subtype_of(db, *existing_negative) {
+                    *self = Self::new();
+                    return;
+                }
+                // A & ~B = A    if A and B are disjoint
+                if existing_negative.is_disjoint_from(db, new_positive) {
+                    to_remove = Some(index);
```

Comment:
This makes sense. This task is also still open.

> One implementation thought: presuming this suggests using a vector for `to_remove`, we could consider `smallvec` backed by a length-1 array to avoid heap allocation in the typical 0-or-1 cases.

:+1:

Reply:
I added this now.